### PR TITLE
Fixes #19009: cis technique compiler fixes

### PR DESCRIPTION
**Pull request by gpoblon**
https://issues.rudder.io/issues/19009

### CODE REVIEW

**Comment by peckpeck** [comment]
is it still work in progress ?

File: rudder-lang/src/generator/cfengine.rs [hunk 1]
```diff
@@ -137,9 +137,24 @@ impl CFEngine {
     fn format_statement(
         &mut self,
         gc: &IR2,
+        res_def: &ResourceDef,
+        state_def: &StateDef,
         st: &Statement,
         in_class: String,
     ) -> Result<Vec<Promise>> {
+        // get variables to try to get the proper parameter value
```
Comment (gpoblon) [inline rudder-lang/src/generator/cfengine.rs:145]: load variables to later replace their value in parameters

File: rudder-lang/src/generator/cfengine.rs [hunk 2]
```diff
@@ -337,28 +364,50 @@ impl CFEngine {
...
+                        r#"{}{}{}:{}"#,
+                        delim,
+                        x,
+                        delim,
+                        self.value_to_string(y, variables, true)?
+                    )),
                     ","
                 )?
             ),
+            Value::Variable(v) => {
```
Comment (gpoblon) [inline rudder-lang/src/generator/cfengine.rs:395]: here parameters values are replaced by variable content, else warn

File: rudder-lang/src/generator/cfengine/syntax.rs [hunk 3]
```diff
@@ -368,9 +368,17 @@ impl Method {
...
         if self.resource == "variable" && self.state == "string_escaped" {
-            let merged_resource_parameters = self.parameters.join(".");
-            self.parameters = vec![merged_resource_parameters.clone()];
-            self.report_parameter = merged_resource_parameters;
+            let mut parameters = Vec::new();
+            for p in self.parameters.clone() {
+                let param = p
+                    .strip_prefix("\"")
+                    .and_then(|stripped_p| stripped_p.strip_suffix("\""))
+                    .unwrap(); // parameters are quoted so unwrap will work without issue
```
Comment (gpoblon) [inline rudder-lang/src/generator/cfengine/syntax.rs:376]: wasnt properly escaped. Not really satisfying way to deal with the issue but it works

File: rudder-lang/src/generator/dsc.rs [hunk 4]
```diff
@@ -80,29 +83,6 @@ impl DSC {
         self.return_condition = None;
     }
 
-    fn parameter_to_dsc(&self, param: &Value, param_name: &str) -> Result<String> {
```
Comment (gpoblon) [inline rudder-lang/src/generator/dsc.rs:83]: was not called anymore

File: rudder-lang/src/generator/dsc.rs [hunk 5]
```diff
@@ -189,21 +185,29 @@ impl DSC {
...
+                let class_param_index = inner_state_def.class_parameter_index(method_name)?;
                 if parameters.len() < class_param_index {
                     return Err(Error::new(format!(
                         "Class param index is out of bounds for {}",
                         method_name
                     )));
                 }
-                // tmp -1, index of lib generator is off 1
+
+                // EXCEPTION: reunite variable_string_escaped resource parameters that appear to be joined from cfengine side
```
Comment (gpoblon) [inline rudder-lang/src/generator/dsc.rs:201]: deal with variable_string exception

File: rudder-lang/src/generator/dsc.rs [hunk 6]
```diff
@@ -243,21 +247,29 @@ impl DSC {
...
-                let state_def = gc.get_state_def(&sd.resource, &sd.state)?;
-                let class_param_index = state_def.class_parameter_index(method_name)?;
+                let class_param_index = inner_state_def.class_parameter_index(method_name)?;
                 if parameters.len() < class_param_index {
                     return Err(Error::new(format!(
                         "Class param index is out of bounds for {}",
                         method_name
                     )));
                 }
+
```
Comment (gpoblon) [inline rudder-lang/src/generator/dsc.rs:261]: deal with variable_string exception (appears again in StateDeclaration)

File: rudder-lang/src/generator/dsc.rs [hunk 7]
```diff
@@ -349,26 +366,74 @@ impl DSC {
...
+    fn value_to_string(
+        &self,
+        value: &Value,
+        variables: &HashMap<&Token, &VariableDef>,
+        string_delim: bool,
+    ) -> Result<String> {
         let delim = if string_delim { "\"" } else { "" };
         Ok(match value {
-            Value::String(s) => format!("{}{}{}", delim, String::try_from(s)?, delim),
+            Value::String(s) => format!(
```
Comment (gpoblon) [inline rudder-lang/src/generator/dsc.rs:377]: Properly escape double quotes + handle parameters value replacement

File: rudder-lang/src/io/output.rs [hunk 8]
```diff
@@ -185,30 +186,36 @@ impl LogOutput {
...
-            } else if files[0].content.is_some() {
-                // no error, expected length = 1
-                println!("{}", &files[0].clone().content.unwrap());
             } else {
                 panic!("BUG! Output should be stdout but there is no content to print");
             }
         }
         // else
         // note there might be no fle to print. ie technique generate
+        let mut filemap: HashMap<&PathBuf, String> = HashMap::new();
```
Comment (gpoblon) [inline rudder-lang/src/io/output.rs:195]: fix : output was only generating 1 technique by file, now it appends properly

File: rudder-lang/src/ir/ir1.rs [hunk 9]
```diff
@@ -180,23 +180,17 @@ impl<'src> IR1<'src> {
...
-                        }
-                    },
-                },
-            }
+            self.errors.push(err!(
+                variable.name,
+                "Global variable definitions are not currently supported ({})",
+                variable.name
+            ));
+            // match VariableDef::from_pvariable_definition(variable, context, &self.enum_list) {
```
Comment (gpoblon) [inline rudder-lang/src/ir/ir1.rs:188]: commented until we decide to make use of global variable definitions again

File: rudder-lang/src/ir/value.rs [hunk 10]
```diff
@@ -161,24 +171,44 @@ impl<'src> Value<'src> {
...
-                map_results(s.data.iter(), |e| match e {
-                    PInterpolatedElement::Static(_) => Ok(()),
-                    PInterpolatedElement::Variable(_v) => Ok(()),
-                    // TODO
-                    //                            match getter(Token::new("", v)) {
-                    //                                None => fail!(s.pos, "Variable {} does not exist at {}", v, s.pos),
-                    //                                _ => Ok(()),
-                    //                            },
-                })
+        let exist = |tk| -> Result<()> {
```
Comment (gpoblon) [inline rudder-lang/src/ir/value.rs:174]: leaving this unimplemented! disallowed usage of different types of variables in rudderlang although they are properly translated into strings. Floats/Int can't hld interpolated values so I dont think any context check is important here

File: rudder-lang/src/ir/variable.rs [hunk 11]
```diff
@@ -66,7 +66,16 @@ impl<'src> CondVariableDef<'src> {
             state_params,
         } = var;
         let (error, metadata) = create_metadata(metadata);
-        context.add_variable_declaration(name, Type::Enum(Token::new("corelib.rl", "boolean")))?;
+        context.add_variable_declaration(name, Type::Enum(Token::new("rudderlang", "boolean")))?;
+        // also add report_data to variables since these are generated and therefore their existence is checked at compile time
+        context.add_variable_declaration(
```
Comment (gpoblon) [inline rudder-lang/src/ir/variable.rs:71]: create a variable for special cases (condition_from). Without these rudderc would return an error with undefined variable calls

**Comment by amousset** [comment]
Done in https://github.com/Normation/rudder/pull/3609

**Pull request closed without merging**